### PR TITLE
Port of AParticularUser's PM Snake moveset & various nerfs

### PR DESCRIPTION
**Pull request by H3ra-57**
Assets:
[hdr-assets-snake.zip](https://github.com/HDR-Development/HewDraw-Remix/files/11668539/hdr-assets-snake.zip)
 

Huge thanks to AParticularUser for allowing us to use their mod in HDR!

# New Features

- side-smash changed to 3 hit knife combo
- down-smash changed to spinning double kick
- c4 will no longer fall off an opponent through knockback, but will unstick itself after 20 seconds
- side-special changed to tranq-gun
- holding left or right during grab will make snake walk
    - pressing attack while walking will make snake do his normal side throws
- up-taunt changed to "infinite ammo" quote
- side-taunt changed to codec-call
- holding left or right during down-taunt will make snake move while in his box
- holding neutral-special during down-taunt will make snake spawn a grenade while in his box
- pressing down-special during down-taunt will make snake place a c4 while in his box
    - pressing down-special again will make snake exit the box and detonate the c4
- Also adds the changes from https://github.com/HDR-Development/HewDraw-Remix/pull/1553

# Changelog

<==Snake==>
FTILT 2
~[+] KBG: 95 --> 100
DASH ATTACK
Late Hit
~[-] Active Frame Window: 8-17 --> 8-14
FSMASH
~[R] Changed to a 3-hit knife combo
Hit 1 (F13/F13/F14 ~ F17)
~ Damage: 6%
~ Angle: 361/170/361
~ BKB: 20/25/30
~ KBG: 25/20/15
~ Hit 2 can be inputted between F17 ~ F34
Hit 2 (F9 ~ F13 after button press 2)
~ Damage: 7%
~ Angle: 361
~ BKB: 20
~ KBG: 25
~ Hit 3 can be inputted between F13 ~ F34
Hit 3 (F9 ~ F13 after button press 3)
~ Damage: 7.5%
~ Angle: 361
~ BKB: 85
~ KBG: 75
DSMASH
~[R] Given a new animation with longer range
Hit 1 (F13 ~ F15)
~ Damage: 12%
~ Angle: 25
~ BKB: 30
~ KBG: 87
Hit 2 (F28 ~ F30)
~ Damage: 14%
~ Angle: 20
~ BKB: 30
~ KBG: 88
NAIR
~[-] Landing lag: 8 --> 12
~[R] FKB: 55 --> 50
Angle change on multihit hitboxes from 367 to 365
Removal of -7 frame hitstun modifiers on multihits
Final Hit
~[-] KBG: 145 --> 120
FAIR
~[-] Landing lag: 13 --> 14
BAIR
Strong Hit
~[R] Angle: 361 --> 50
~[-] KBG: 100 --> 95
Weak Hit
~[-] Damage: 10 --> 5
~[-] Active Frames: 16 --> 12
~[R] Angle: 361 --> 45
UPAIR
Strong Hit (Feet)
~[-] KBG: 100 --> 90
~[+] BKB: 35 --> 45
Strong Hit (Legs)
~[-] KBG: 100 --> 70
Weak Hit
~[-] KBG: 100 --> 80
DAIR
~[-] Landing lag: 10 --> 12
GRAB
~[R] Snake will now walk when holding left or right on the control stick, pressing the attack button while walking will perform the respective throw
~[/] New walking while grabbing animation added
GRENADE
~[-] Snake exclusive hitbox normalized with frame 2
~[-] Hitbox Size: 20 --> 16
NIKITA → TRANQUILIZER
~(!) Changes Nikita to Tranquilizer, a gunshot that put the opponent to sleep
~ Bullet is shot on F24
~ Damage: 1%
C4
~[+] Sticky will no longer fall off from taking damage.
~[-] Sticky will fall off after 20 seconds.
UTAUNT, STAUNT
~(!) New taunts added
DTAUNT
~[+] Snake can now stay in the box for a maximum of 20 seconds
~[+] Can now move during this taunt by holding left and right
~[+] Can now place grenades and C4 during this taunt
~[+] Can now detonate C4 during this taunt, doing so will end the taunt
~ C4 detonated this way will explode on F50

### CODE REVIEW

**Comment by SuddyN** [comment]
he has the thing

https://github.com/HDR-Development/HewDraw-Remix/assets/33226440/866558e7-b91d-49e8-9e5a-0e17d2089059

**Review by SuddyN — CHANGES_REQUESTED** [review]
changelog is missing some things that are included in the Snake Nerfs PR, such as NAir angle change. It would be easier to review and for the sake of Askew if this changelog included everything

File: fighters/snake/src/acmd/aerials.rs [hunk 1]
```diff
@@ -21,11 +21,11 @@ unsafe fn snake_attack_air_n_game(fighter: &mut L2CAgentBase) {
...
-        ATTACK(fighter, 4, 0, Hash40::new("hip"), 2.0, 367, 100, 55, 0, 4.0, 0.0, 0.0, 0.0, None, None, None, 1.0, 1.0, *ATTACK_SETOFF_KIND_ON, *ATTACK_LR_CHECK_F, true, 0, 0.0, 0, false, false, false, false, true, *COLLISION_SITUATION_MASK_A, *COLLISION_CATEGORY_MASK_ALL, *COLLISION_PART_MASK_ALL, false, Hash40::new("collision_attr_normal"), *ATTACK_SOUND_LEVEL_S, *COLLISION_SOUND_ATTR_KICK, *ATTACK_REGION_KICK);
+        ATTACK(fighter, 4, 0, Hash40::new("hip"), 2.0, 365, 100, 55, 0, 4.0, 0.0, 0.0, 0.0, None, None, None, 1.0, 1.0, *ATTACK_SETOFF_KIND_ON, *ATTACK_LR_CHECK_F, true, 0, 0.0, 0, false, false, false, false, true, *COLLISION_SITUATION_MASK_A, *COLLISION_CATEGORY_MASK_ALL, *COLLISION_PART_MASK_ALL, false, Hash40::new("collision_attr_normal"), *ATTACK_SOUND_LEVEL_S, *COLLISION_SOUND_ATTR_KICK, *ATTACK_REGION_KICK);
```
Comment (SuddyN) [inline fighters/snake/src/acmd/aerials.rs]: FKB was not decreased for this hit

**Review by SuddyN — DISMISSED** [review]
approved everything except status

File: fighters/snake/src/status.rs [hunk 2]
```diff
@@ -0,0 +1,737 @@
+use super::*;
+use globals::*;
+use smashline::*;
+
```
Comment (blu-dev) [inline fighters/snake/src/status.rs:4]: Needs more comments

**Review by blu-dev — CHANGES_REQUESTED** [review]

File: fighters/snake/src/status.rs [hunk 3]
```diff
@@ -0,0 +1,737 @@
...
+#[status_script(agent = "snake_c4", status = WEAPON_SNAKE_C4_STATUS_KIND_ESTABLISH_TARGET, condition = LUA_SCRIPT_STATUS_FUNC_STATUS_PRE)]
+unsafe fn snake_c4_status_pre(weapon: &mut L2CFighterCommon) -> L2CValue {
+    StatusModule::init_settings(
+        weapon.module_accessor,
+        SituationKind(*SITUATION_KIND_AIR),
+        *WEAPON_KINETIC_TYPE_NONE,
+        *GROUND_CORRECT_KIND_AIR as u32,
+        GroundCliffCheckKind(0),
+        false,
+        *WEAPON_STATUS_WORK_KEEP_FLAG_SNAKE_C4_DROP_FALL_FLAG, // <-- wrong flag make it check ground collision?
```
Comment (blu-dev) [inline fighters/snake/src/status.rs:56]: ?

File: fighters/snake/src/status.rs [hunk 4]
```diff
@@ -0,0 +1,737 @@
...
+        *WEAPON_STATUS_WORK_KEEP_FLAG_SNAKE_C4_ESTABLISH_TARGET_FLOAT,
+        0,
+    );
+    0.into()
+}
+
+
+#[status_script(agent = "snake", status = FIGHTER_SNAKE_STATUS_KIND_SPECIAL_LW_PRODUCE, condition = LUA_SCRIPT_STATUS_FUNC_STATUS_MAIN)]
+unsafe fn snake_down_special_start_status_main(fighter: &mut L2CFighterCommon) -> L2CValue {
+    if WorkModule::is_flag(fighter.module_accessor, *FIGHTER_SNAKE_STATUS_SPECIAL_LW_FLAG_SQUAT) {
```
Comment (blu-dev) [inline fighters/snake/src/status.rs]: I'd prefer if all of these were `fighter.is_flag` and associated getters/setters, `fighter.module_accessor` gets old.

File: fighters/snake/src/status.rs [hunk 5]
```diff
@@ -0,0 +1,737 @@
...
+            fighter.change_status(FIGHTER_SNAKE_STATUS_KIND_SPECIAL_LW_ESTABLISH_JUDGE.into(), false.into());
+        }
+        false.into()
+    }
+    else {
+        fun_7100018800(fighter, false);
+        true.into()
+    }
+}
+pub unsafe fn fun_7100018800(fighter: &mut L2CFighterCommon, skip_check: bool) {
```
Comment (blu-dev) [inline fighters/snake/src/status.rs]: Find a name for this based on it's behavior

File: fighters/snake/src/status.rs [hunk 6]
```diff
@@ -0,0 +1,737 @@
...
+    if skip_check == false {
+        if StatusModule::situation_kind(fighter.module_accessor) == StatusModule::prev_situation_kind(fighter.module_accessor) {
+            return
+        }
+    }
+    if StatusModule::situation_kind(fighter.module_accessor) == *SITUATION_KIND_GROUND {
+        GroundModule::set_rhombus_offset(fighter.module_accessor, &Vector2f{x:0.0, y:0.0});
```
Comment (blu-dev) [inline fighters/snake/src/status.rs]: &Vector2f::zero()

File: fighters/snake/src/status.rs [hunk 7]
```diff
@@ -0,0 +1,737 @@
...
+        KineticModule::change_kinetic(fighter.module_accessor, *FIGHTER_KINETIC_TYPE_GROUND_STOP);
+        fighter.set_situation(SITUATION_KIND_GROUND.into());
```
Comment (blu-dev) [inline fighters/snake/src/status.rs:124]: *SITUATION_KIND_GROUND instead of .into()

File: fighters/snake/src/status.rs [hunk 8]
```diff
@@ -0,0 +1,737 @@
...
+        fighter.set_situation(SITUATION_KIND_AIR.into());
+        GroundModule::correct(fighter.module_accessor, GroundCorrectKind(*GROUND_CORRECT_KIND_AIR));
+        let motion = WorkModule::get_int64(fighter.module_accessor, *FIGHTER_SNAKE_STATUS_WORK_INT_MOT_AIR_KIND);
+        MotionModule::change_motion_inherit_frame(fighter.module_accessor, Hash40::new_raw(motion), -1.0, 1.0, 0.0, false, false);
+    }
+}
+
+#[status_script(agent = "snake", status = FIGHTER_SNAKE_STATUS_KIND_SPECIAL_LW_PRODUCE, condition = LUA_SCRIPT_STATUS_FUNC_STATUS_END)]
+unsafe fn snake_down_special_start_status_end(fighter: &mut L2CFighterCommon) -> L2CValue {
+    if fighter.global_table[0xb].get_i32() != *FIGHTER_SNAKE_STATUS_KIND_SPECIAL_LW_ESTABLISH_GROUND
```
Comment (blu-dev) [inline fighters/snake/src/status.rs]: We have actual names for the constants, this one is probably `PREV_STATUS_KIND`, you can check `dynamic/src/constants.rs`

File: fighters/snake/src/status.rs [hunk 9]
```diff
@@ -0,0 +1,737 @@
...
+        fighter.change_status(FIGHTER_STATUS_KIND_FALL.into(), false.into());
+        // true.into()
+    }
+    true.into()
+}
+
+// added self-stick-squat and back-air to down-special check
+#[status_script(agent = "snake", status = FIGHTER_STATUS_KIND_SPECIAL_LW, condition = LUA_SCRIPT_STATUS_FUNC_STATUS_MAIN)]
+unsafe fn snake_down_special_status_main(fighter: &mut L2CFighterCommon) -> L2CValue {
+    if fighter.global_table[10].get_i32() == *FIGHTER_STATUS_KIND_SQUAT_WAIT
```
Comment (blu-dev) [inline fighters/snake/src/status.rs]: Same thing as above.

File: fighters/snake/src/status.rs [hunk 10]
```diff
@@ -0,0 +1,737 @@
...
+    ControlModule::reset_trigger(fighter.module_accessor);
+    MotionModule::change_motion(fighter.module_accessor, Hash40::new("catch_pull"), 0.0, 1.0, false, 0.0, false, false);
+    fighter.sub_shift_status_main(L2CValue::Ptr(snake_grab_pull_main_loop as *const () as _))
+    // 0.into()
+}
+#[status_script(agent = "snake", status = FIGHTER_STATUS_KIND_CATCH_DASH_PULL, condition = LUA_SCRIPT_STATUS_FUNC_STATUS_MAIN)]
+unsafe fn snake_grab_dash_pull_status_main(fighter: &mut L2CFighterCommon) -> L2CValue {
+    ControlModule::reset_trigger(fighter.module_accessor);
+    MotionModule::change_motion(fighter.module_accessor, Hash40::new("catch_pull"), 0.0, 1.0, false, 0.0, false, false);
+    fighter.sub_shift_status_main(L2CValue::Ptr(snake_grab_pull_main_loop as *const () as _))
```
Comment (blu-dev) [inline fighters/snake/src/status.rs:352]: Everywhere this is written you can do `fighter.main_shift` instead

File: fighters/snake/src/status.rs [hunk 11]
```diff
@@ -0,0 +1,737 @@
...
+    else if MotionModule::is_end(fighter.module_accessor) {
+        if PostureModule::lr(fighter.module_accessor)*ControlModule::get_stick_x(fighter.module_accessor) > 0.1
+        || PostureModule::lr(fighter.module_accessor)*ControlModule::get_stick_x(fighter.module_accessor) < -0.1{
+            VarModule::on_flag(fighter.object(), vars::snake::instance::IS_GRAB_WALK);
+        }
+        fighter.change_status(FIGHTER_STATUS_KIND_CATCH_WAIT.into(), false.into());
+        return true.into()
+    }
+    else if MotionModule::frame(fighter.module_accessor) >= 1.0 {
+        if fighter.global_table[0x21].get_i32() & *FIGHTER_PAD_CMD_CAT2_FLAG_THROW_LW != 0 {
```
Comment (blu-dev) [inline fighters/snake/src/status.rs]: Ditto on the constant

File: fighters/snake/src/status.rs [hunk 12]
```diff
@@ -0,0 +1,737 @@
...
+                fighter.change_status(FIGHTER_STATUS_KIND_THROW.into(), false.into());
+                return true.into()
+            }
+            else{
+                VarModule::on_flag(fighter.object(), vars::snake::instance::IS_GRAB_WALK);
+                fighter.change_status(FIGHTER_STATUS_KIND_CATCH_WAIT.into(), false.into());
+                return true.into()
+            }
+        }
+        else if PostureModule::lr(fighter.module_accessor)*ControlModule::get_stick_x(fighter.module_accessor) < -0.7 {
```
Comment (blu-dev) [inline fighters/snake/src/status.rs]: add some whitespace here pls

File: fighters/snake/src/status.rs [hunk 13]
```diff
@@ -0,0 +1,737 @@
...
+    fighter.sub_shift_status_main(L2CValue::Ptr(snake_grab_wait_main_loop as *const () as _))
+    // 0.into()
+}
+pub unsafe fn snake_grab_wait_main_loop(fighter: &mut L2CFighterCommon) -> L2CValue {
+    let entry_id = WorkModule::get_int(fighter.module_accessor, *FIGHTER_INSTANCE_WORK_ID_INT_ENTRY_ID) as usize;
+    if StatusModule::situation_kind(fighter.module_accessor) == *SITUATION_KIND_AIR {
+        fighter.change_status(FIGHTER_STATUS_KIND_CATCH_CUT.into(), false.into());
+        return true.into()
+    }
+    else if fighter.global_table[0x21].get_i32() & *FIGHTER_PAD_CMD_CAT2_FLAG_THROW_LW != 0 {
```
Comment (blu-dev) [inline fighters/snake/src/status.rs]: Actually, use `fighter.is_cat_flag` please here instead.

**Review by blu-dev — DISMISSED** [review]

File: fighters/snake/src/status.rs (hunk 3 above)
Comment (H3ra-57) [reply fighters/snake/src/status.rs:56]: your guess is as good as mine

File: fighters/snake/src/status.rs (hunk 7 above)
Comment (H3ra-57) [reply fighters/snake/src/status.rs:124]: this caused compile errors, and it just suggested i do the old thing
image

File: fighters/snake/src/status.rs (hunk 10 above)
Comment (H3ra-57) [reply fighters/snake/src/status.rs:352]: this caused compile errors (mismatched type; `main_shift` wants ptrs but the input is not a pointer)

**Review by SkewedAskew — APPROVED** [review]

**Pull request merged**